### PR TITLE
Versioning : Adding more resilience

### PR DESCRIPTION
**Pull request by mihirbhende-okta**
While I have other RFC PR to simplify versioning here https://github.com/askspoke/spoke/pull/30313, that proposed a much bigger change were risk/reward is much lower. 

It made me think that can be do `insert a new document but get the version from other document and do + 1 before we save` all on DB level. After lot of coffee, I was able to do this using mongodb's `$merge` aggregation operator https://www.mongodb.com/docs/manual/reference/operator/aggregation/merge/

I have added comments to explain this in detail. But the idea is `never relay on version which is in memory`.

This is much smaller change and should ideally work without any changes in unit tests. In fact I have a unit test exactly for this in main repo where we assert than no matter how many concurrent operations happen, we should always have resiliency between version and its shadows.

I have added tests for the same as PR in main repo : https://github.com/askspoke/spoke/pull/30325

### CODE REVIEW

**Comment by ankitgoyal-okta** [comment]
@mihirbhende-okta 
current version - n - List has values A, B, C
2 updates simultaneously came

n + 1 - remove value B, add value D.
n + 1 - remove value B, add value E.
So what would be the current version in the collection and what would be the last version in shadow collection.

If the 1. is processed first then the current version will have value A, C, E.

If the 2. is processed first then the current version will have value A, C, D.

So this is still true right?

**Comment by mihirbhende-okta** [comment]
> @mihirbhende-okta current version - n - List has values A, B, C 2 updates simultaneously came
> 
> n + 1 - remove value B, add value D. n + 1 - remove value B, add value E. So what would be the current version in the collection and what would be the last version in shadow collection.
> 
> If the 1. is processed first then the current version will have value A, C, E.
> 
> If the 2. is processed first then the current version will have value A, C, D.
> 
> So this is still true right?

Yes, you can find example in here : [askspoke/spoke#30325](https://github.com/askspoke/spoke/pull/30325)

**Comment by ankitgoyal-okta** [comment]
> > @mihirbhende-okta current version - n - List has values A, B, C 2 updates simultaneously came
> > n + 1 - remove value B, add value D. n + 1 - remove value B, add value E. So what would be the current version in the collection and what would be the last version in shadow collection.
> > If the 1. is processed first then the current version will have value A, C, E.
> > If the 2. is processed first then the current version will have value A, C, D.
> > So this is still true right?
> 
> Yes, you can find example in here : [askspoke/spoke#30325](https://github.com/askspoke/spoke/pull/30325)

You get it why we can't do this, user A who made the change will see the end state different then what he expects and similary user B. Also they are changing the original document not the modified one. Remember we talked about this few months back and that's why we never made this change. If you want more context I can explain it on a call.

**Comment by mihirbhende-okta** [comment]
> You get it why we can't do this, user A who made the change will see the end state different then what he expects and similary user B. Also they are changing the original document not the modified one. Remember we talked about this few months back and that's why we never made this change. If you want more context I can explain it on a call.

Yeah, this is expected because we don't have locking. The window for this happen this is micro seconds i.e. the execution time of the aggregation and DB save. So I would say probability is very very low. Currently we are in much worse state because it gets the version from memory and it's easy to come into situation where it's stale. Currently if this happens, we throw `versions do not match` in the same case.

**Comment by mihirbhende-okta** [comment]
> Also they are changing the original document not the modified one

Thats one point of concurrent processing which we can run into in general because we do not have locking or data refresh in UI. The main objective of this change is to minimize the edge case window, and if we run into it.. Then handle that gracefully so upstream functions, specially in data sync do not fail.

**Comment by ankitgoyal-okta** [comment]
> > You get it why we can't do this, user A who made the change will see the end state different then what he expects and similary user B. Also they are changing the original document not the modified one. Remember we talked about this few months back and that's why we never made this change. If you want more context I can explain it on a call.
> 
> Yeah, this is expected because we don't have locking. The window for this happen this is micro seconds i.e. the execution time of the aggregation and DB save. So I would say probability is very very low. Currently we are in much worse state because it gets the version from memory and it's easy to come into situation where it's stale. Currently if this happens, we throw `versions do not match` in the same case.

We do have locking, versions do not match works as a locking mechanism. We are doing this change exactly for the reason that concurrent changes are happening otherwise why we are making this change. The reason they are not handling concurrent changes is that the base version changes so the user is not getting what they are expecting and the end state would depend on how we process the changes instead of how user performs it.

**Comment by mihirbhende-okta** [comment]
> user is not getting what they are expecting

Considering the current behavior : the locking will throw. So the user B who is in UI gets an error as user A saved it first. But we do not update the UI in that case that `hey someone else updated it just before you did so we locked and failed, here is the updated configlist`. The user has to refresh anyways to get the updated configlist.

The main concern here is user A thought he updated the coniglist to XXX state but it was saved to YYY state. Along with new approach here, we can get around that user experience by doing `configlist.findMyself` before we return response. This way the save response will update the UI and user will know that list is into YYY state. And then they can update it again if they want to?

**Comment by mihirbhende-okta** [comment]
On a side note, thank you @ankitgoyal-okta for asking these questions 😃. Versioning is definitely a tricky part and I would love to get it as close to perfection as possible while handling concurrency so it helps us in long run. We are having lot more admins doing the updates now in lists, so we run into version error more frequently than before.

**Comment by ankitgoyal-okta** [comment]
> > user is not getting what they are expecting
> 
> Considering the current behavior : the locking will throw. So the user B who is in UI gets an error as user A saved it first. But we do not update the UI in that case that `hey someone else updated it just before you did so we locked and failed, here is the updated configlist`. The user has to refresh anyways to get the updated configlist.
> 
> The main concern here is user A thought he updated the coniglist to XXX state but it was saved to YYY state. Along with new approach here, we can get around that user experience by doing `configlist.findMyself` before we return response. This way the save response will update the UI and user will know that list is into YYY state. And then they can update it again if they want to?

I still think throwing is better than showing them something totally unexpected.

In the second thing if you are showing user updated state and want user to make the change again on the updated list then you don't need all this right. In the UI if the error is thrown just refresh the page and you are done.

**Comment by mihirbhende-okta** [comment]
> In the second thing if you are showing user updated state and want user to make the change again on the updated list then you don't need all this right. In the UI if the error is thrown just refresh the page and you are done.

Expecting users to refresh the page to handle locking is definitely risky. I definitely get your concern here but I would rather tell them that `hey someone updated it before you did, so this is what we have in DB now. You can either keep it or update it again to what you wanted`. Rather than `hey we failed`. Also most of the users do not consider refresh as an option and they will keep trying. 

I forgot to mention another case which is not entirely related configlist updates from U. We have seen that admins run manual sync from settings UI and then data sync job runs via cron at the same time. The lists for which saves happen with a race condition, the versioning throws error to lock. As a result, the main sync for that external resource fails. This is where the new change will be handy as well. I would love to know what @zachhart-okta thinks as he just spent some time working around that 😁

**Comment by ankitgoyal-okta** [comment]
> a result, the main sync for that external res

If we are making this change then any change that has been made would not lead to a definitive state either by hand or by data sync job and that is my biggest problem with this change and other than this I don't know what other utility we are getting.

**Comment by mihirbhende-okta** [comment]
> If we are making this change then any change that has been made would not lead to a definitive state either by hand or by data sync job and that is my biggest problem with this change and other than this I don't know what other utility we are getting.

Handling concurrency without an error is the main utility here 😆. By throwing an error, I think we are just avoiding that operation and adding overhead later to get things back in state. We recently ran few migrations to fix versioning so data sync is unblocked. Avoid that was the main purpose behind this change.

`The last operation in race was committed and has correct version associate with it` is a definitive new stage I am proposing. I think it comes down to the design choice we will make i.e. `fail the second save operation` vs `let the second operation win the race and gets saved overriding earlier version`. I am down to not implementing this if this is adding more issues than solving the core problem.

But from what I see, this will allow us to complete the syncs and also show the user that document has been updated when we refetch the saved document before returning in response. If a user A has to add his changes again because user B did some updates just before him, I think its reasonable expectation than failing the save or having versions out of sync. 😉

**Comment by macyabbey-okta** [comment]
Gracefully handling concurrent client operations can be achieved with `If-Match` and/or `If-Modified-Since` http headers. This is the classic `click save, UX tells you there have been changes, would you like to reload?`

This is an orthogonal concern, which should be addressed separately.

Both user experience options, `throw` and `ta-da, you got back results from soneone else's update` are not good options, so trading one for the other is not a big difference in `pro`/`con` for me.

The big win here is the narrowing of edge case time window, and decreasing seemingly non-deterministic behavior of sync and versioning.

Big win!!

**Review by macyabbey-okta — APPROVED** [review]
🚀 🎸

**Comment by zachhart-okta** [comment]
The biggest issue with the current functionality is this scenario, which happens across two saves of a list:

1. the `pre('save')` hook in this plugin runs successfully, clones the "main" document and saves it to the shadow collection with `_version` `n`.
2. The _actual_ `save()` call on the main document, now with version `n+1` in memory, **fails** - probably due to a _mongoose version error_ (not the versioning errors we get from this library, which are duplicate `_id` errors, but errors against mongoose's `_v` property), or some other concurrency issue. This step may be happening at the same time as 3.
3. Another process comes around and attempts to save the document, the `pre('save')` hook in this plugin runs again. This time, it cannot save a copy of the document to the shadow collection _no matter what_ because the main document in the database _still has a version `n`_, this plugin has no method of recovering from this case.
4. Since the `pre('save')` hook fails, the `save()` call itself fails. We're in a cycle that cannot be recovered from except for manual intervention from us. 

This plugin needs to derive version numbers from something other than the in memory copy of the document, or needs to have some sort of collision resolution when it runs into the case where it would try to write a document to the shadow collection with a duplicate `_id`. The way the plugin is working currently puts us in scenarios that the plugin itself cannot recover from. Without some additional locking mechanism in order to prevent steps 2 and 3 as described above from running concurrently, this plugin will be the source of more errors as it successfully writes a shadow document but fails to update the main document. 

There may be other solutions here too, rather than querying the main document before versioning. Perhaps some [error handling middleware](https://mongoosejs.com/docs/5.x/docs/middleware.html#error-handling-middleware) that actually _removes_ the shadow document that would prevent future save calls from failing?

**Comment by zachhart-okta** [comment]
Would creating a `configlists.locks` collection that just holds a document whose `_id` is the `_id` of the configlist we're writing and truly just having an actual, queryable lock , adding pre('save') and post('save') hooks to the configlist model to set (create) and release (delete) the lock on _every_ write solve this just as well, without getting into the weeds with versioning? If we can just prevent the race condition from happening at all by not allowing this plugin to run concurrently with itself, I think we'd solve our problems. It certainly adds a bit of overhead to write operation on the configlists, and we'd have to think through implications for things like datasync (datasync operations should probably take the lock out at the start of the sync, not pre-save, but datasync is already special so it can use mongo driver methods I guess to skip the mongoose hooks), but it's conceptually very simple

File: source/versioning.js [hunk 1]
```diff
@@ -158,58 +159,53 @@ module.exports = function (schema, options) {
...
+                        [constants.VERSION]: '$_version'
+                    },
+                    [constants.VALIDITY]: {
+                        start: `\$${[constants.VALIDITY].start}`,
+                        end: now
+                    },
+                    [constants.VERSION]: '$_version'
+                }
+            },
+            { $merge: { into: shadowCollection, on: '_id' , whenMatched: "replace", whenNotMatched: "insert"  } }
```
Comment (zachhart-okta) [inline source/versioning.js:196]: Might be the first `$merge` aggregation step in the codebase, exiting times.

**Review by zachhart-okta — APPROVED** [review]
My biggest motivator here in getting some new behavior out is what I mentioned above - this plugin can, today, put our data in a state where we can't recover short of running a migration. To add on top of that, all the existing complexities in datasync and partial failures/successess there (far more likely to have a user run into the datasync than have two users run into each other, imo) make repeated failures due to this duplicate `_id` failure state more costly to fix. 

With relationships between documents across collections now (configListIds stored on ExternalResources), it's going to become more and more critical that these complicated sync and document relationships be as reliable as possible. I think narrowing the time window to enter this unrecoverable state from 100s of ms into the single ms or ns range is a huge win for us. 

We're in a spot now where the integrity of the configlist and external resource collections is in the critical path for granting access to resources - if we're plagued by data integrity issues here it will reflect poorly on spoke and IGA, and that has some real business impact.

**Comment by suchitagarwal-okta** [comment]
Great solution here Mihir! 

to Zach's point above, as far as i can tell, we have to have locks to guarantee these operations are atomic

**Comment by mihirbhende-okta** [comment]
Bump @kylepeterson-okta @willpotter-okta

**Review by kylepeterson-okta — APPROVED** [review]

**Pull request merged**